app/models/access_control_upload.rb

```diff
@@ -0,0 +1,463 @@
+      group.update(user_ids: user_ids)
+    end
+  end
+
+  private def import_collections!
+    collections.each do |item|
+      collection = Collection.where(name: item[:name], collection_type: collection_type(item)).first_or_create!
+      collection_relations.each_key do |relation|
+        names = item[relation].select { |m| m[:found] }.map { |r| r[:name] }
+        case relation
+        when :coc_codes
+          collection.update(coc_codes: names) if names.present?
+        else
+          ids = collection.class_name_for_viewable_type(relation).constantize.where(name: names).pluck(:id)
+          collection.set_viewables({ relation => ids }) if ids.present?
+        end
+      end
+    end
+  end
+
+  private def import_access_controls!
+    # Make sure we know about roles, collections, and user groups we created earlier
+    clear_cache
+    access_controls.select { |acl| acl[:existing_user_id].blank? }.each do |item|
+      next if item[:error]
+
+      role = existing_role(item[:role])
+      collection = existing_collection(item[:collection])
+      user_group = existing_user_group(item[:user_group])
+      AccessControl.create!(role: role, collection: collection, user_group: user_group)
+    end
+  end
+
+  private def titles
+    @titles ||= [].tap do |h|
+      Role.permissions(exclude_health: true).count.times do |i|
+        title = roles_worksheet[1][2 + i]&.value
+        h << title if title.present?
+      end
+    end
+  end
+
+  private def permissions_for_row(row, existing_role)
+    {}.tap do |p|
+      Role.permissions(exclude_health: true).count.times do |i|
+        title = titles[i]
+        value = case row[2 + i]&.value
+        when 'X', 'x', 'Y', 'y', 'TRUE', 'true'
+          true
+        else
+          false
+        end
+        perm = Role.permission_from_title(title)
+        next unless perm
+
+        p[perm[:column]] = {
+          title: title,
+          existing_value: existing_role.try(:[], perm[:column]),
+          incoming_value: value,
+        }
+      end
+    end
+  end
+
+  private def workbook
+    @workbook ||= ::RubyXL::Parser.parse_buffer(file.download)
+  end
+
+  private def roles_worksheet
+    workbook['Roles']
+  end
+
+  private def users_worksheet
+    workbook['Users']
+  end
+
+  private def collections_worksheet
+    workbook['Collections']
+  end
+
+  private def valid_roles?
+    roles_worksheet.cell_at('A2').value == 'Role' && roles_worksheet.cell_at('B1').value == 'Permission'
+  end
+
+  private def valid_users?
+    users_worksheet.cell_at('A1').value == 'First Name' && users_worksheet.cell_at('B1').value == 'Last Name'
+  end
+
+  private def valid_collections?
+    collections_worksheet.cell_at('A1').value == 'Collection' && collections_worksheet.cell_at('B1').value == 'Data Sources'
+  end
+
+  private def existing_role(role_name)
+    existing_roles[role_name]
+  end
+
+  private def existing_roles
+    @existing_roles ||= Role.editable.index_by(&:name)
```

Comment:
fwiw, not sure that role has a unique constraint on name (or even a rails validation). Don't need to fix that but maybe just order them so the behavior is reproducible (same on these collections and user_groups)
```suggestion
    @existing_roles ||= Role.editable.order(:id).index_by(&:name)
```

Reply:
Yeah, that definitely has occurred to me.  Project doesn't either, but we're using it as a primary key for importing collections.  We'll have to see if that becomes an issue.